### PR TITLE
Add EIP: Interface of Confidential Transactions Supported Token Contract

### PR DESCRIPTION
**Pull request by andrewcoder666**
This PR proposal draws up a standard interface of confidential transaction supported token contracts, by providing basic functionality without loss of generality. Contracts following the standard can provide confidentiality for users’ balances and token transfer value.

### CODE REVIEW

File: ERCS/erc-7945.md [hunk 1]
```diff
@@ -0,0 +1,250 @@
+---
+eip: 7945
+title: Confidential Transactions Supported Token
+description: General interface for confidential transaction supported token contract with allowance and approve functionality.
```
Comment (SamWilsn) [inline ERCS/erc-7945.md]: > General interface for confidential transaction supported token contract

All of the above is either fluff ("general interface") or restating your title ("confidential transaction supported token"). You should try to expand on the ideas introduced in your title.

You might consider answering questions like "what is a confidential transaction?", "is this for fungible or non-fungible tokens?", and/or "is this an extension of an existing standard?"

File: ERCS/erc-7945.md [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+status: Draft
+type: Standards Track
+category: ERC
+created: 2025-05-09
+requires: 20
+---
+
+## Abstract
+
+This proposal draws up a standard interface of confidential transaction supported token contracts, by providing basic functionality without loss of generality. Contracts following the standard can provide confidentiality for users’ balances and token transfer value.
```
Comment (SamWilsn) [inline ERCS/erc-7945.md]: This is more like an introduction than a complete abstract. You should include a bit about the goals of the proposal, and a technical overview of how the proposal accomplishes those goals.

File: ERCS/erc-7945.md [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+## Motivation
+
+A standard interface allows confidential transactions of tokens on Ethereum (and/or other EVM-compatible blockchains) to be applied by certain parties which are sensitive to transfer amount, or by privacy-preserving applications.
+
```
Comment (SamWilsn) [inline ERCS/erc-7945.md:29]: The motivation section is where you convince the reader of two things: (1) that the goals of the proposal are worth achieving, and (2) that this particular proposal is the best approach to reaching those goals.

In this case, for (1) you should discuss the need for confidential transactions. You could perhaps include some use cases that would benefit from confidential transactions. For (2), you want to compare this particular proposal against other (potentially hypothetical) proposals and explain how this one is the best option available.

File: ERCS/erc-7945.md [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+
+### Contract Interface
+
+#### Methods
+
+##### name
```
Comment (SamWilsn) [inline ERCS/erc-7945.md]: Here, and in the rest of the document, please surround inline code with backticks (\`):

```suggestion
##### `name`
```

File: ERCS/erc-7945.md [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+  * The plaintext `value` behind `_confidentialTransferValue` is non-negative;
+  * The caller's confidential balance is actually enough to pay the plaintext `value` behind `_confidentialTransferValue` .
+
+```Solidity
+function confidentialTransfer(
+  address _to,
+  bytes memory _confidentialTransferValue, 
+  bytes memory _proof
+) public returns (bool success)
+```
```
Comment (SamWilsn) [inline ERCS/erc-7945.md:119]: You may want to move the signature to the top of the section so the reader has context for the variable names.

File: ERCS/erc-7945.md [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+##### confidentialTransfer
+
+Transfers `value` amount of tokens (behind `_confidentialTransferValue` ) to address `_to` , and MUST fire the `ConfidentialTransfer` event. The function SHOULD `throw` if the message caller’s `_proof` of this transfer fails to be verified.
+
+Note:
+
+* Callers MUST handle `false` from returns (bool success). Callers MUST NOT assume that `false` is never returned!
+* Implementations can fully customize the proof system, (de)serialization strategies of `bytes` and/or the business workflow. For example, when implementing "Zether" (DOI: 10.1007/978-3-030-51280-4_23) confidential token contracts, the `_confidentialTransferValue` and accounts' confidential balances will be encrypted homomorphically under ElGamal public keys, and `_proof` will consist 3 parts to check:
```
Comment (SamWilsn) [inline ERCS/erc-7945.md]: Please use the CSL-JSON format [defined in EIP-1](https://eips.ethereum.org/EIPS/eip-1#digital-object-identifier-system) for citations using a DOI.

File: ERCS/erc-7945.md [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+```
+
+## Rationale
+
+
+Confidential transactions have been implemented in many blockchains, either natively through blockchain protocols like Monero and Zcash, or through smart contracts like Zether (DOI: 10.1007/978-3-030-51280-4_23) without modifying blockchain protocol.
+
+However, when it comes to the latter way, actually no standards are proposed yet to illustrate such contracts. Users and applications cannot easily detect whether a token contract supports confidential transactions or not, and so hardly can they make transfers without revealing the actual amount.
+
+Consequently, this proposal is to standardize the confidential transaction supported token contracts, meanwhile without loss of generality, by only specifying core methods and events.
```
Comment (SamWilsn) [inline ERCS/erc-7945.md]: This content is describing the proposal as a whole, and so belongs in the rationale section. My pet example for the distinction between motivation and rationale is:

> Motivation: We need to build a shed because...
> Rationale: We chose to paint the shed red because...

File: ERCS/erc-7945.md [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+
+To support audit of confidential transactions and total supply, especially such token issuers are banks or other financial institutes supervised by governments or monetary authorities, confidential transactions can be implemented without changing `confidentialTransfer` method signature, by encoding more info into parameters.
+
+For example in Zether-like implementation (DOI: 10.1007/978-3-030-58951-6_29), if token transfers are required to be audited, `confidentialTransfer` caller encrypts transfer `value` redundantly under public keys of caller's, `to`'s, a group of auditors', which makes it possible that related parties can exactly know the real `value` behind. So does `confidentialTotalSupply()`.
+
+### Fat Token
+
+Confidential transactions supported token can also implement [ERC-20](./erc-20.md) at the same time.
+
+Token accounts in such tokens can hold 2 kinds of balances. Such token contracts can optionally provide methods to hind [ERC-20](./erc-20.md) plaintext balances into confidential balances, and vice versa, to reveal confidential balances back to [ERC-20](./erc-20.md) plaintext balances.
```
Comment (SamWilsn) [inline ERCS/erc-7945.md]:
```suggestion
Token accounts in such tokens can hold 2 kinds of balances. Such token contracts can optionally provide methods to hide [ERC-20](./erc-20.md) plaintext balances into confidential balances, and vice versa, to reveal confidential balances back to [ERC-20](./erc-20.md) plaintext balances.
```

Perhaps?

File: assets/erc-7945/value_equation.png
Comment (SamWilsn) [inline assets/erc-7945/value_equation.png:1]: We do support inline LaTeX using [MathJax](https://www.mathjax.org/). If that is insufficient, we have a preference for SVGs if you can provide one. If not, PNG is acceptable.

File: ERCS/erc-7945.md (hunk 2 above)
Comment (wizzelman801) [reply ERCS/erc-7945.md]: trying to catch up with my billing for my adds and its putting me in a hole to start my business.

File: ERCS/erc-7945.md (hunk 3 above)
Comment (wizzelman801) [reply ERCS/erc-7945.md:29]: this would help me transition my life around and my bussiness ventures.

File: ERCS/erc-7945.md (hunk 4 above)
Comment (wizzelman801) [reply ERCS/erc-7945.md]: Sherwin Manalon

File: ERCS/erc-7945.md (hunk 5 above)
Comment (wizzelman801) [reply ERCS/erc-7945.md:119]: Sherwin Manalon

Comment (wizzelman801) [reply ERCS/erc-7945.md:119]: Sherwin  Manalon

File: ERCS/erc-7945.md (hunk 1 above)
Comment (andrewcoder666) [reply ERCS/erc-7945.md]: updated.

File: ERCS/erc-7945.md (hunk 2 above)
Comment (andrewcoder666) [reply ERCS/erc-7945.md]: > This is more like an introduction than a complete abstract. You should include a bit about the goals of the proposal, and a technical overview of how the proposal accomplishes those goals.

updated.

File: ERCS/erc-7945.md (hunk 8 above)
Comment (andrewcoder666) [reply ERCS/erc-7945.md]: > Perhaps?

typo fixed

File: ERCS/erc-7945.md (hunk 6 above)
Comment (andrewcoder666) [reply ERCS/erc-7945.md]: > Please use the CSL-JSON format [defined in EIP-1](https://eips.ethereum.org/EIPS/eip-1#digital-object-identifier-system) for citations using a DOI.

DOI format issue fixed

File: ERCS/erc-7945.md (hunk 5 above)
Comment (andrewcoder666) [reply ERCS/erc-7945.md:119]: done

File: ERCS/erc-7945.md (hunk 4 above)
Comment (andrewcoder666) [reply ERCS/erc-7945.md]: done

File: ERCS/erc-7945.md (hunk 3 above)
Comment (andrewcoder666) [reply ERCS/erc-7945.md:29]: updated

File: ERCS/erc-7945.md (hunk 7 above)
Comment (andrewcoder666) [reply ERCS/erc-7945.md]: moved to motivation

**Comment by andrewcoder666** [comment]
sorry to bother you guys @g11tech @SamWilsn @xinbenlv . can you help check with this pull request again?

File: ERCS/erc-7945.md [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+status: Draft
+type: Standards Track
+category: ERC
+created: 2025-05-09
+requires: 20
+---
+
+## Abstract
+
+Classic token contracts like [ERC-20](./eip-20.md) enable their token holders to make transfers and/or approve others to make transfers on behalves. Generality of token standard [ERC-20](./eip-20.md) catalyzed decentralizing finance and many other blockchain applications. However, when it comes to the privacy, as its emergence of importance, although some technical schemes get proposed, few standards have been established, which limits the evolution of privacy-preserving blockchain applications.
```
Comment (SamWilsn) [inline ERCS/erc-7945.md]:
```suggestion
Classic token contracts like [ERC-20](./eip-20.md) enable their token holders to make transfers and/or approve others to make transfers on their behalves. The generality of token standard [ERC-20](./eip-20.md) catalyzed decentralizing finance and many other blockchain applications. However, when it comes to privacy, although some technical schemes have been proposed, few standards have been established, which limits the evolution of privacy-preserving blockchain applications.
```

Here are a few grammar fixes, but you'll probably want to go over this again and make sure it's well worded.

File: ERCS/erc-7945.md [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+
+This proposal draws up a standard interface of confidential transaction supported fungible token contracts, by providing basic transfer functionality without loss of generality as well as providing allowance and approve functionalities, so contracts following the standard can provide confidentiality for users’ balances and token transfer value, and can enable other blockchain applications to make transfers on behalves, which empowers more privacy-preserving capabilities to blockchain applications.
```
Comment (SamWilsn) [inline ERCS/erc-7945.md]:
```suggestion
This proposal draws up a standard interface for fungible token contracts supporting confidential transactions. It provides basic transfer functionality without loss of generality, and allowance and approve functionalities. Contracts following the standard can provide confidentiality for users’ balances and token transfer value, and can enable other blockchain applications to make transfers on behalf of owners, which empowers more privacy-preserving capabilities to blockchain applications.
```

Again, here's a few suggestions for grammar fixes, but you'll want to double check.

File: ERCS/erc-7945.md [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+        }
+      ],
+      "DOI": "10.1007/978-3-030-58951-6_29",
+      "title": "PGC: Decentralized Confidential Payment System with Auditability",
+      "original-date": {
+        "date-parts": [
+          [2020, 9, 12]
+        ]
+      },
+      "URL": "https://link.springer.com/chapter/10.1007/978-3-030-58951-6_29"
```
Comment (SamWilsn) [inline ERCS/erc-7945.md]: The `URL` field _must_ point to a freely accessible copy of the paper, while the `additional-urls` property can link to paid resources (see [EIP-1](https://eips.ethereum.org/EIPS/eip-1#digital-object-identifier-system) for an example).

File: ERCS/erc-7945.md [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+        }
+      ],
+      "DOI": "10.1007/978-3-030-51280-4_23",
+      "title": "Zether: Towards Privacy in a Smart Contract World",
+      "original-date": {
+        "date-parts": [
+          [2020, 2, 10]
+        ]
+      },
+      "URL": "https://dl.acm.org/doi/abs/10.1007/978-3-030-51280-4_23"
```
Comment (SamWilsn) [inline ERCS/erc-7945.md]: Same comment here. We need to be able to read the linked content for free. If there's a way to get the content from this link, I didn't see it :sweat_smile:

File: ERCS/erc-7945.md [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+
+Such standard interface allows confidential transactions of tokens to be applied by certain parties which are sensitive to transfer amount, or by privacy-preserving applications.
+
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+
+### Contract Interface
```
Comment (SamWilsn) [inline ERCS/erc-7945.md:36]: This looks loosely patterned on the ERC-20 style of writing interfaces. While there's nothing wrong with that, modern proposals generally include one or more Solidity interfaces. This makes it a bit easier to directly use in contracts, and has a better chance of maintaining consistency.

In this case, you could do one interface for `name`, `symbol`, and `decimals`; and another for everything else. This has additional benefits for ERC-165 identification.

File: ERCS/erc-7945.md [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+
+OPTIONAL - This method can be used to improve usability, but interfaces and other contracts MUST NOT expect these values to be present.
+
+##### `confidentialBalanceOf`
+
+```Solidity
+function confidentialBalanceOf(address owner) 
+public view returns (bytes memory confidentialBalance)
+```
+Returns the account confidential balance of another account with address `owner` .
```
Comment (SamWilsn) [inline ERCS/erc-7945.md]: What does "another" here mean?

**Review by SamWilsn — CHANGES_REQUESTED** [review]

File: ERCS/erc-7945.md (hunk 14 above)
Comment (andrewcoder666) [reply ERCS/erc-7945.md]: fixed. now: "Returns the confidential balance of the account with address `owner` ."

File: ERCS/erc-7945.md (hunk 13 above)
Comment (andrewcoder666) [reply ERCS/erc-7945.md:36]: fixed.  name, symbol, and decimals are extracted to ITokenMetadata. other functions are extracted to IConfidentialToken.

File: ERCS/erc-7945.md (hunk 12 above)
Comment (andrewcoder666) [reply ERCS/erc-7945.md]: change url to open-access one.

File: ERCS/erc-7945.md (hunk 11 above)
Comment (andrewcoder666) [reply ERCS/erc-7945.md]: change url to open-access one.

**Comment by andrewcoder666** [comment]
Hi @SamWilsn  , I updated the document based on your comments. Could you please review this ERC proposal again?

File: ERCS/erc-7945.md [hunk 15]
```diff
@@ -0,0 +1,342 @@
+---
+eip: 7945
+title: Confidential Transactions Supported Token
+description: Confidential transactions hide token amounts. This proposal outlines a fungible confidential token interface, boosting privacy in dApps.
```
Comment (SamWilsn) [inline ERCS/erc-7945.md:4]:
```suggestion
description: Outlines a fungible token interface that hides per-account balances, boosting privacy in dApps
```

Perhaps a bit more refinement? What you have is pretty reasonable though.

**Comment by SamWilsn** [comment]
Cerberus, block wizzelman801

**Review by SamWilsn — APPROVED** [review]
Looks good! Just this optional comment.
